### PR TITLE
Initial implementation of c2c calls

### PR DESCRIPTION
**Pull request by wkaluza**
Adds ledger support for contracts like the following:

```
// caller contract
contract c1_interface
  @action
  function eleven() : Int64;
endcontract

@action
function c2c_call(contract_name : String)
  contract c = c1_interface(contract_name);

  assert(c.eleven() == 11i64);
endfunction
```
```
// callee contract
@action
function eleven() : Int64
  return 11i64;
endfunction
```

### CODE REVIEW

File: scripts/install-test-dependencies.sh [hunk 1]
```diff
@@ -1,6 +1,7 @@
 #!/bin/bash
 set -e
 
-pip3 install  wheel pyyaml requests --force-reinstall
-pip3 install  fetchai-ledger-api==0.11.0a3 --force-reinstall
+pip3 install wheel --force-reinstall
+pip3 install pyyaml requests --force-reinstall
```
Comment (wkaluza) [inline scripts/install-test-dependencies.sh:5]: pyyaml needs wheel to install without errors, but pip is not clever enough to know that.

File: libs/ledger/src/chaincode/smart_contract.cpp [hunk 2]
```diff
@@ -495,8 +498,119 @@ Contract::Result SmartContract::InvokeAction(std::string const &name, chain::Tra
...
+      ss << "Execution of function " << function.name << " from contract " << identity
+         << " failed with error \"" << error << "\"";
+      error = ss.str();
+    }
+
+    c.state_adapter->PopContext();
+    vm->IncreaseChargeTotal(vm2.GetChargeTotal() - reference_charge);
+
+    auto it = call_history.find(called_contract_address);
+    call_history.erase(it);
```
Comment (ejfitzgerald) [inline libs/ledger/src/chaincode/smart_contract.cpp]: Think that it is probably safer (and more concise) in the following form:

```suggestion
    call_history.erase(called_contract_address);
```

File: libs/ledger/src/chaincode/smart_contract.cpp [hunk 3]
```diff
@@ -495,8 +498,119 @@ Contract::Result SmartContract::InvokeAction(std::string const &name, chain::Tra
...
+      error = "Contract-to-contract call cycle detected";
+
+      return false;
+    }
+
+    call_history.insert(called_contract_address);
+
+    decltype(auto) c = context();
+
+    // TODO(WK) charge for reading from storage
```
Comment (ejfitzgerald) [inline libs/ledger/src/chaincode/smart_contract.cpp]: Can we raise an issue for this one

**Review by ejfitzgerald — DISMISSED** [review]
Really excellent work, just a couple of very minor comments

**Review by ejfitzgerald — DISMISSED** [review]

**Review by ejfitzgerald — APPROVED** [review]

**Pull request merged**